_posts/2020-09-30-hows-the-performance-of-ruby-3.0.0-preview1.md

```diff
@@ -0,0 +1,80 @@
+
+## Results
+
+How are results? Not dramatic, I'm afraid.
+
+<table>
+    <thead>
+        <tr><td></td><th>Ruby 2.7</th><th>Ruby 3.0.0-preview1</th><th>Speedup/Slowdown</th></tr>
+    </thead>
+    <tbody>
+        <tr>
+            <th>Median Throughput</th><td>165.7</td><td>160.7</td><td style="color:red">-3.0%</td>
+            <th>Fastest Run of 30</th><td>168.7</td><td>164.0</td><td style="color:red">-2.8%</td>
+            <th>Slowest Run of 30</th><td>163.0</td><td>158.2</td><td style="color:red">-2.9%</td>
+        </tr>
+    </tbody>
+</table>
+
+(I also have [the raw data](http://codefol.io/links/ruby_3.0.0-preview1_rrb_results.json.gz), if that's of use to you. And technically there's an even more raw, larger set of results. [Hit me up on Twitter](https://twitter.com/codefolio) if you want them.)
+
+To summarize simply: they're nearly the same speed, and a bit a pre-release Ruby polish and/or fixes to my testing will probably get them back to exactly equal.
+
+## Other Issues...
+
+Keep in mind that preview1 isn't final. I'm going to investigate this and see what I can see. My code branch is certainly not perfect. We're seeing ***ugly*** interactions between the now-ancient Discourse code I'm using and Ruby 3.0's deprecations. And I may find some improvements along the way.
+
+Still, what I'm seeing suggests that these results aren't far off. A severe bug wouldn't cost a few percent of performance - it would kill it, or boost it to utterly unreasonable levels. That's not what we're seeing here. (I'm going to see if I can find where Rails and/or my hackery slows things down, of course.)
+
+This is also a very limited test. I could check Ruby versions (again) much farther back, [though you can see the previous results of that](https://engineering.appfolio.com/appfolio-engineering/2019/3/7/ruby-speed-roundup-20-through-26), and they shouldn't change. [Ruby 2.6 and 2.7 are functionally the same speed to within 1%-2%](https://engineering.appfolio.com/appfolio-engineering/2019/12/27/ruby-270s-rails-ruby-bench-speed-is-unchanged-from-260). It looks like 2.7 and the 3.0 preview aren't showing a significant speed boost for Rails.
+
+Is that shocking? It shouldn't be. I was, frankly, very surprised to see the 72%-ish speed boost from 2.0 to 2.6. Rails spends a lot of its time I/O-bound, waiting on databases, files and the network. The current Ruby JIT champions, JRuby and TruffleRuby can't easily squeeze more performance out of Rails than CRuby in most cases. The garbage collector, a source of slowdown back in Ruby 2.0 and 2.1, runs extremely solidly for these use cases.
+
+I think we're seeing about the speed that Rails is going to be, for Ruby 3.0 and for some time afterward.
+
+With that said, I think a combination of polishing Ruby 3.0 for release and me making sure my test is in order will return the few percentage points of speed that (my tests claim) Ruby 3.0 has lost versus 2.6 and 2.7.
```

Comment:
It would be great to link to one of our articles at the end if you are okay with that. Maybe something like "If you are interested in performance and speech checkout some more fastruby articles on the [subject](https://www.fastruby.io/blog/performance/rails/writing-fast-rails.html).

Reply:
Sounds good! I've added a link to the performance tag on FastRuby.io, which seems appropriate.